### PR TITLE
WIP: Baby zkSNARKS soundness proof formalisation

### PR DESCRIPTION
**Pull request by DanielRrr**

### CODE REVIEW

File: ZkSNARK/GeneralLemmas/MvXMul.lean [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+universe u v
+variable {R : Type u} 
+
+namespace MvPolynomial
+variable {σ : Type _} {e : ℕ} {n m : σ} {s : σ →₀ ℕ}
+
+section CommSemiringₓ
+
+variable [CommSemiringₓ R] {p q : MvPolynomial σ R}
+variable [CommSemiring R]
```
Comment (winston-h-zhang) [inline ZkSNARK/GeneralLemmas/MvXMul.lean]: Hmm, this should probably not be here -- we already have a `[CommSemiringₓ R]`. Adding this one
1. is not backward compatible with mathlib3
2. might cause a bunch of weird instance conflicts in the future

Comment (DanielRrr) [reply ZkSNARK/GeneralLemmas/MvXMul.lean]: The last two lemmas don't work without [CommSemiring R] for some reason

Comment (DanielRrr) [reply ZkSNARK/GeneralLemmas/MvXMul.lean]: Lean says that it failed to synthesise the CommSemiring instance

Comment (winston-h-zhang) [reply ZkSNARK/GeneralLemmas/MvXMul.lean]: That probably means you should change the lemmas; it's just that adding `[CommSemiring R]` causes conflicts -- when we use those lemmas with mathlib3 theorems, they will have no idea what a `CommSemiring` is. For them only `CommSemiringₓ` exists.

Comment (DanielRrr) [reply ZkSNARK/GeneralLemmas/MvXMul.lean]: They are proved using the `ring` tactic

Comment (DanielRrr) [reply ZkSNARK/GeneralLemmas/MvXMul.lean]: We may leave this for a while

```
lemma add_mul_distrib [CommSemiring R] (a b c d : R) : a + b * c + b * d = a + b * (c + d) :=
by ring

lemma add_mul_distrib' [CommSemiring R] (a b c d : R) : a + c * b + d * b = a + b * (c + d) :=
by ring
```

This works as well

**Review by winston-h-zhang — APPROVED** [review]

**Pull request merged**